M2/Macaulay2/packages/Varieties/SheafMaps.m2

```diff
@@ -0,0 +1,890 @@
+-- Prune
+-----------------------------------------------------------------------------
+
+-- Consider the sequence 0 -> m^[p] -> S -> S/m^[p] -> 0 and apply Hom(-,M)
+minimalPresentation SheafMap :=
+prune SheafMap := SheafMap => opts -> f -> (
+    (G, F) := (target f, source f);
+    prune G; prune F; -- these are pruned just to populate cached data
+    -- F.cache.TorsionFree = M/H^0_B(M)
+    g := inducedMap(G.cache.TorsionFree, truncate(degree f, F.cache.TorsionFree), matrix f);
+    p := max(G.cache.GlobalSectionLimit, F.cache.GlobalSectionLimit);
+    -- TODO: substitute with appropriate irrelevant ideal
+    Bp := module (ideal vars ring F)^[p];
+    lift sheafMap prune Hom(Bp, g))
+
+-----------------------------------------------------------------------------
+-- Things to move to the Core
+-----------------------------------------------------------------------------
+
+random Module := Vector => o -> M -> vector(gens M * random(cover M, module ring M, o))
+homomorphism Vector := v -> homomorphism matrix v
+
+-----------------------------------------------------------------------------
+-- pullback and pushout
+-----------------------------------------------------------------------------
+
+-- TODO: also for a list of matrices
+pullback(SheafMap, SheafMap) := CoherentSheaf => {} >> o -> (f, g) -> (
+    -- TODO: use != instead
+    if target f =!= target g then error "expected maps with the same target";
+    h := map(target f, source f ++ source g, matrix f | -matrix g);
+    P := ker h;
+    S := source h;
+    P.cache.pullbackMaps = {
+	S^[0] * inducedMap(S, P),
+	S^[1] * inducedMap(S, P)};
+    P)
+
+-- TODO: also for a list of matrices
+pushout(SheafMap, SheafMap) := CoherentSheaf => (f, g) -> (
+    -- TODO: use != here instead
+    if source matrix f =!= source matrix g then error "expected maps with the same source";
+    h := map(target f ++ target g, source f, matrix f || -matrix g);
+    P := coker h;
+    T := target h;
+    P.cache.pushoutMaps = {
+	inducedMap(P, T) * T_[0],
+	inducedMap(P, T) * T_[1]};
+    P)
+
+trans := (C,v) -> (
+    if C.cache.?indexComponents then (
+	    Ci := C.cache.indexComponents;
+	    apply(v, i -> if Ci#?i then Ci#i else error "expected an index of a component of the direct sum"))
+    else (
+        if not C.cache.?components then error "expected a direct sum of coherent sheaves";
+	    Cc := C.cache.components;
+	    apply(v, i -> if not Cc#?i then error "expected an index of a component of the direct sum");
+	    v)
+    )
+
+CoherentSheaf _ Array := SheafMap => (F, v) -> (
+    v = trans(F,v);
+    G := directSum apply(toList v, j -> F.cache.components#j);
+    map(F, G, (cover module F)_v))
+
+CoherentSheaf ^ Array := SheafMap => (F, v) -> (
+    v = trans(F,v);
+    G := directSum apply(toList v, j -> F.cache.components#j);
+    map(G, F, (cover module F)^v))
+
+- SheafMap := phi -> map(target phi, source phi, -matrix phi)
+
```

Comment:
Should we allow multiplication by any element of the underlying ring?  What about addition when the source and targets agree?